Add database

uncle-sly opened the pull request:
PR Sprint 12

u-sly committed:
Sprint 12, solution.
InDbUserRepository ready and checked by Postman tests.

u-sly committed:
Sprint 12, solution.
Добавлены контроллеры, сервисы, репозитории для таблиц схемы.

u-sly committed:
UserFriends переименован UserFriend, а так же объекты с ним связанные.

u-sly committed:
Реализована обработка фильмов, наполнение рейтином и жанрами. Добавлен метод для batchUpdate в BaseRepository

u-sly committed:
Доработана обработака Фильмов, добавлен exception, исправлен синтаксис в запросах, исправлено наполнение таблиц в data.sql
Postman тесты пройдены успешно.

u-sly committed:
Доработана возможность добавляться в друзья с подтверждением

u-sly committed:
переименование 2 классов

u-sly committed:
методы
addFriend,deleteFriend,getFriends,getCommonFriends
вынесены из UserRepository и InDbUserRepository в UserFriendRepository

все тесты postman прошли.

u-sly committed:
рефакторинг названий классов и структуры проекта

u-sly committed:
добавлены тесты репозиториев.

u-sly committed:
исправление в тестах, вместо instance of на Objects::NoNull

u-sly committed:
исправление Checkstyle

u-sly committed:
исправление Checkstyle

u-sly committed:
исправление Checkstyle

KirPyt commented on pom.xml:
```diff
@@ -41,6 +41,29 @@
 			<artifactId>logbook-spring-boot-starter</artifactId>
 			<version>3.7.2</version>
 		</dependency>
+<!--		12 sprint -->
```
Советую не комментировать части кода для прошлой функциональности, а делать отдельный коммит с поясняющим комментарием, так проще версионировать и отслеживать изменения

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/FilmorateApplication.java:
```diff
@@ -7,6 +7,8 @@
 public class FilmorateApplication {
 	public static void main(String[] args) {
 		SpringApplication.run(FilmorateApplication.class, args);
+		System.out.println("Current working directory: " + System.getProperty("user.dir"));
```
Советую использовать логгер для справочной информации

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandlingControllerAdvice.java:
```diff
@@ -63,10 +64,19 @@ public ErrorResponse onValidationException(final ValidationException e) {
+
     @ExceptionHandler
     @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
     public ErrorResponse onException(final Exception e) {
```
Здесь стоит перейти к перехватыванию класса исключений `Throwable`, как наиболее общего класса исключений. Это позволит отловить любую исключительную ситуацию на сервере

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/repository/UserRepository.java:
```diff
@@ -16,12 +15,12 @@ public interface UserRepository {
     User update(User user);
 
-    void addFriend(User user, User friend);
+//    void addFriend(User user, User friend, boolean isConfirmed);
```
Все закомментированные методы убираем. Удаление методов и причину можно отразить в коммите

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/repository/inDatabase/InDbBaseRepository.java:
```diff
@@ -0,0 +1,101 @@
+        }, keyHolder);
+
+        Long id = null;
+//      System.out.println(keyHolder.getKeys());
```
Все закомментированный код убираем. Удаление кода и причину можно отразить в коммите

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/repository/inDatabase/InDbBaseRepository.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    protected int[] batchInsert(String query, Long filmId, ArrayList<Genre> params) {
+        return jdbc.batchUpdate(query,
```
Отличное решение использовать пакетное обновление

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/repository/inDatabase/InDbUserRepository.java:
```diff
@@ -0,0 +1,98 @@
+    private static final String INSERT_QUERY = "INSERT INTO users (email, login, name, birthday)" +
+            "VALUES (?, ?, ?, ?)";
+    private static final String UPDATE_QUERY = "UPDATE users SET email = ?, login = ?, name = ?, birthday = ? WHERE id = ?";
+//    private static final String INSERT_FRIEND_QUERY = "INSERT INTO user_friend (user_Id, friend_id, isConfirmed) VALUES (?, ?, ?)";
```
Аналогично, закомментированный код убираем

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/repository/inDatabase/InDbUserRepository.java:
```diff
@@ -0,0 +1,98 @@
+        return user;
+    }
+
+//    public void addFriend(User user, User friend, boolean isConfirmed) {
```
Аналогично, закомментированный код убираем

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -16,10 +22,27 @@ public class FilmServiceImpl implements FilmService {
+        if (null != film.getGenres()) {
+
+            film.setGenres(new LinkedHashSet<>(film.getGenres().stream()
+                    .map(id -> genreRepository.getById(id.getId())
```
Здесь происходит циклическое обращение к БД, этого следует избегать, так как сильно нагружает базу. Стоит постараться передать список требуемых жанров и сразу вернуть требуемый вид данных из БД

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -31,9 +54,19 @@ public Film update(final Film film) {
+                    .orElseThrow(() -> new FkConstraintViolationException("Рейтинг вне диапазона."))
+            );
+        film.setGenres(new HashSet<>(filmGenreRepository.getById(film.getId()).stream()
+                    .map(id -> genreRepository.getById(id.getGenreId())
```
Здесь происходит циклическое обращение к БД, этого следует избегать, так как сильно нагружает базу. Стоит постараться передать список требуемых жанров и сразу вернуть требуемый вид данных из БД

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserServiceImpl.java:
```diff
@@ -57,15 +62,21 @@ public void deleteFriend(long userId, long friendId) {
+                .collect(Collectors.toSet());
+
+        return userRepository.getAll().stream()
+                .filter(u -> uf.contains(u.getId()))
```
Стоит постараться реализовать логику фильтрации за счёт sql запроса к базе (получить требуемый результат, но за счёт выполнения запроса к БД)

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserServiceImpl.java:
```diff
@@ -76,7 +87,17 @@ public List<User> getCommonFriends(long userId, long otherId) {
+                .map(Map.Entry::getKey) // Извлекаем friendId
+                .collect(Collectors.toSet()); // Собираем в Set
+
+        return userRepository.getAll().stream()
```
Стоит постараться реализовать логику фильтрации за счёт sql запроса к базе (получить требуемый результат, но за счёт выполнения запроса к БД)

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserServiceImpl.java:
```diff
@@ -76,7 +87,17 @@ public List<User> getCommonFriends(long userId, long otherId) {
                 .orElseThrow(() -> new ValidationException("Пользователь c id: " + otherId + " не найден"));
 
-        return userRepository.getCommonFriends(user.getId(), other.getId());
+        Set<Long> uf = userFriendRepository.getCommonFriends(user.getId(), other.getId()).stream()
```
Рекомендую попробовать реализовать отбор общих друзей за счёт формирования запроса к базе данных (sql запрос), передав идентификаторы требуемых юзеров

KirPyt reviewed (commented):
Здравствуйте!

Посмотрел Ваш проект и указал некоторые комментарии и замечания по коду. В целом мне понравилась идея с реализацией слоя работы с базой данных (адресация запросов дополнительному компоненту, который уже передает в базу), это уменьшает связанность в коде и делает его удобным для поддержки

Если будут вопросы по ревью, то пишите в Пачку

u-sly committed:
удалил аннотацию AutoConfigureTestDatabase т.к. она уже входит в @JdbcTest

u-sly committed:
чистка pom.xml, замена на logger в FilmorateApplication

u-sly committed:
добавлен перехват класса исключений Throwable

u-sly committed:
удален закоментированный код
в метод insert добавлена проверка Objects.requireNonNull вместо if

u-sly committed:
UserRepository интерфейс:
удалены т.к. вынесены другой репозиторий addFriend и deleteFriend.
Добавлены getFriends и getCommonFriends.

InDbUserFriendRepository
удалены методы getFriends и getCommonFriends и связаннные запросы, т.к. они относятся к сущности пользователь.
Переименованы методы добавления и удаления записи о дружбе.

InDbUserRepository
удалены методы addFriend и deleteFriend, т.к. вынесены в репозиторий InDbUserFriendRepository.
вернул сюда getFriends и getCommonFriends.

UserServiceImpl
в методах getFriends и getCommonFriends, убрана логика обработки данных репозиториев, вместо этого вызываются методы репозиториев в к-х логика реализована с помощью sql запросов. Данные возвращаются уже подготовленные для отдачи в контроллер.

Удалены:
InDbUserFriendRepositoryTests.java
Не проверяем отдельно таблицу связей, проверять будем сущность User.
UserFriendRowMapper.java
UserFriend.java

Пройдены все тесты Postman.

u-sly committed:
Genre и Mpa: добавлены конструкторы.

FilmExtractor - добавлен новый класс, для обработки resultset фильма.

FilmRowMapper - добавлено сохранение имени рейтинга Mpa.

FilmServiceImpl:
метод getById логика получения фильма вынесена в репозиторий и в sql
метод save - так же вынесена в репозиторий, в слой работы с данными.

InDbFilmRepository:
sql запросы изменены, вынесены из констант свой соответствующий метод.
метод get - получение фильма переписан с использованием  filmExtractor.
метод getAll - переписан, получаю данные из 3 репозиториях и собираю в методе, чтобы все фильмы вернуть с заполненным рейтингом и жанрами.
метод save - добавлено сохранение жанров.

u-sly committed:
Film и Genre
добавлен EqualsAndHashCode по id

InDbGenreRepository
добавлен метод getByIds получения Жанров по входящему списку id жанров.

InDbFilmGenreRepository
добавлен метод delete для удаления связей Film - Genre.

InDbFilmRepository
метод update - запрос sql вынесен из константы в метод, добавлено удаление связей film-genre перед добавлением новых.

FilmServiceImpl
удален закоментареный код
метод update - добавлена проверка на валидность рейтинга и жанров входящего фильма.

u-sly committed:
FilmServiceImpl
метод save - добавлена проверка null для жанров входящего фильма

FilmExtractor
добавлена проверка жанров, для обработки фильмов без жанров.

Film и Genre
удалена аннотация EqualsAndHashCode

InDbFilmRepository
изменен запрос для метода get, добавлен left join, чтобы находить фильмы без жанров.

В остальных репозиториях запроса из констант перенесены в соответствующие методы.

u-sly committed:
FilmGenre
удалена модель т.к. это не сущность а таблица связей в бд

FilmGenreRowMapper
перенес на уровень InDbFilmRepository, изменен для использования только с внутренней структурой InDbFilmRepository, т.е. для record FilmGenre.

InDbFilmGenreRepository
изменен в связи с удалением модели
оставлены необходимые методы для работы с таблицей бд, остальные удалены.

InDbFilmRepository
rowmapper для рекорда вынесен в отдельный класс и добавлен сюда через конструктор.

На данный коммит все тесты Postman пройдены успешно.

u-sly committed:
Тестовое окружение: добавлены application.properties и test-data.sql

Добавлены и исправлены тесты. Все пройдены успешно.

u-sly committed:
Добавлены интерфейсы MpaRepository и GenreRepository.
В сервисах заменены ссылки на интерфейсы репозиториев, вместо реализации.
Классы переименованы в Jdbc...., т.к. в данном задании в реализации используем JDBC.

Все тесты Postman пройдены успешно.

u-sly committed:
JDBC - замена JdbcTemplate на NamedParameterJdbcOperations
Т.к. JdbcTemplate это реализация, а наследоваться следует не от конкретной реализации, а от интерфейса, она заменена на интерфейс NamedParameterJdbcOperations.
Соответственно переписан базовый класс JdbcBaseRepository.java и все остальные классы репозиториев на использование NamedParameterJdbcOperations.
Т.к. NamedParameterJdbcOperations позволяет использовать именнованные параметры в запросах, то переписаны все sql запросы в методах. Добавлены Map для передачи параметров sql запросов.

Все тесты Postman пройдены успешно.
Все интеграционные тесты пройдены успешно.

u-sly committed:
Удален закоментированный код.
Все тесты Postman пройдены успешно.
Все интеграционные тесты пройдены успешно.

u-sly committed:
Исправление ошибок Checkstyle.

u-sly committed:
spring.datasource.url=jdbc:h2:mem:filmorate

u-sly committed:
правка pom для dependency com.h2database

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/repository/inDatabase/JdbcFilmRepository.java:
```diff
@@ -0,0 +1,124 @@
+
+        List<Genre> genres = genreRepository.getAll();
+        List<Film> films = findMany(sql, Collections.emptyMap());
+        List<FilmGenre> filmGenres = jdbc.query(sql2, filmGenreRowMapper);
```
Рекомендую давать для переменных читаемые названия, это позволит облегчить понимание кода другими программистами

KirPyt reviewed (commented):
Здравствуйте!

Рад, что получилось поправить комментарии по коду, Ваш проект зачтён, хорошего дня!

uncle-sly commented:
> Здравствуйте!
> 
> Рад, что получилось поправить комментарии по коду, Ваш проект зачтён, хорошего дня!

Спасибо, Кирилл!

Поставьте, пожалуйста, апрув, чтобы я мог смерджить ветку:
"Review required. At least 1 approving review is required by reviewers with write access. "
